pkg/kv/kvserver/kvstorage/storage.go

```diff
@@ -121,8 +126,20 @@ type Engines struct {
 // MakeEngines creates an Engines handle in which both state machine and log
 // engine reside in the same physical engine.
 func MakeEngines(eng storage.Engine) Engines {
-	// TODO(#158281): in test builds, wrap the engines in a way that allows
-	// verifying that all accesses touch the correct subset of keys.
+	if util.RaceEnabled {
+		// Wrap the engines with span set engines to catch incorrect engine
+		// accesses.
+		wrappedStateEngine := spanset.NewSpanSetEngine(eng)
+		wrappedStateEngine.(*spanset.SpanSetEngine).AddForbiddenMatcher(validateIsStateEngineSpan)
```

Comment:
Make `NewSpanSetEngine` return the raw type, instead of casting?

Alternatively, pass the validator into `NewSpanSetEngine`. This wrapper is a no-op without a validator, so one can say it's part of this engine creation.